core/sr-primitives/src/lib.rs

```diff
@@ -374,6 +537,127 @@ impl From<codec::Compact<Perbill>> for Perbill {
 	}
 }
 
+/// Representation of a sign bit.
+///
+/// Note that the ordering matters for proper `Ord` derive.
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+pub enum Sign {
+	/// Negative
+	N,
+	/// Positive
+	P,
+}
+
+/// A fixed point number in the range of `[-(u64::max+1), u64::max+1]`. Composed of one `u64`
+///  _natural_ part and a `Perbill` _decimal_ part.
+///
+/// Note that a positive instance is always greater than a negative instance, even if both have `0`
+///  as their natural and decimal part.
+///
+/// Note that internal Perbill supports an inclusive range of of [0, 1], hence the number actually
+/// be `n` + `perbill::one()`, which is technically `n + 1`.
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+pub struct Fixed64 {
+	/// Sign of the number
+	sign: Sign,
+	/// The natural part.
+	n: u64,
+	/// The decimal part.
+	decimal: Perbill,
+}
+
+#[allow(dead_code)]
+impl Fixed64 {
+	fn positive(&self) -> bool {
+		self.sign == Sign::P
+	}
+
+	fn negative(&self) -> bool {
+		self.sign != Sign::N
+	}
+
+	fn new(sign: Sign, n: u64, decimal: Perbill) -> Self {
+		Self { sign, decimal, n }
+	}
+
+	fn abs(&self) -> Self {
+		Self { sign: Sign::P, n: self.n, decimal: self.decimal }
+	}
+
+	fn from_perbill(decimal: Perbill) -> Self {
+		Self { decimal, ..Default::default() }
+	}
+}
+
+impl ops::Neg for Fixed64 {
+	type Output = Self;
+
+    fn neg(self) -> Self::Output {
+		let mut r = self;
+        match r.sign {
+            Sign::P => r.sign = Sign::N,
+            Sign::N => r.sign = Sign::P,
+        };
+		r
+    }
+}
+
+impl ops::Add for Fixed64 {
+	type Output = Self;
+
+	fn add(self, rhs: Self) -> Self {
+		if self.sign == rhs.sign {
+			// sign will not change. Simply add them ignoring the sign
+			let new_decimal = match self.decimal.checked_add(&rhs.decimal) {
+				Some(d) => d,
+				None => Perbill::from_parts((self.decimal.0 + rhs.decimal.0) % 1_000_000_000)
+			};
+			Self {
+				sign: self.sign,
+				n: self.n.saturating_add(rhs.n), /// TODO
+				decimal: new_decimal
```

Comment:
I think this could be written more quickly with

n = sef.n saturating add rhs.n + (self.decimal.0 + rhs.decimal.0).rem(billion)
deciman = (self.decimal.0 +otherdecimal) % billion